### PR TITLE
Fix non-ascii content garbled in log

### PR DESCRIPTION
**Pull request by junyuecao**
I think decodeURIComponent is the right way to deal with encoding here. The \<result\> here is every byte of the utf-8 string but String.fromCharCode() regard the input as UTF-16

### CODE REVIEW

**Comment by kripken** [comment]
I'm not sure what this change does - perhaps add a comment in the code, and a testcase?

**Comment by junyuecao** [comment]
> I'm not sure what this change does - perhaps add a comment in the code, and a testcase?

@kripken Here is some detail about this change. Please review again and I'll add a testcase and comment in the code.

The result returned from formatString is an array with uint8 that represent the utf-8 content of the string. So to decode the array we can:
1. convert it to a encoded string like `%E4%B8%AD%E6%96%87` and decodeURIComponent
2. use a [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) \(in Node.js a [util.TextDecoder](https://nodejs.org/api/util.html#util_new_textdecoder_encoding_options) \)
3. convert byte array to charcode array



```javascript
// This is the result "中" returned from formatString. It can not convert directly using fromCharCode
var result = [228, 184, 173];
// Current: Using String.fromCharCode(garbled)
ä¸­
// 1. Convert to "%E4%B8%AD" and decodeURIComponent
decodeURIComponent("%E4%B8%AD") ===  "中"
// 2. Using TextDecoder
new TextDecoder().decode(new Uint8Array([228, 184, 173])) === "中"
// 3. More complecated ignore here
```

**Comment by kripken** [comment]
I see, thanks @junyuecao - looks good!

Not sure why the test runs have have been cancelled. Please merge in latest upstream, that should get a new set of tests to pass.

File: src/library.js [hunk 1]
```diff
@@ -4249,11 +4249,8 @@ LibraryManager.library = {
 
   emscripten_log__deps: ['$formatString', 'emscripten_log_js'],
   emscripten_log: function(flags, format, varargs) {
-    var str = '';
     var result = formatString(format, varargs);
-    for (var i = 0 ; i < result.length; ++i) {
-      str += String.fromCharCode(result[i]);
-    }
+    var str = UTF8ArrayToString(result, 0, result.length);
```
Comment (sbc100) [inline src/library.js:4253]: I see most other caller of this function just omit the 3rd argument.  Can you do that here too?

Comment (junyuecao) [reply src/library.js:4253]: OK. That also works. I'll omit it.

**Comment by junyuecao** [comment]
> I see, thanks @junyuecao - looks good!
> 
> Not sure why the test runs have have been cancelled. Please merge in latest upstream, that should get a new set of tests to pass.

Thanks @kripken @sbc100 . Pushed a new commit and the test rerun. Please review again.

**Comment by kripken** [comment]
A test would be nice. `grep` for `emscripten_log` under `tests/` can probably find an existing test that we can just add a new line to.

If there isn't one, I wouldn't block this on that though.

**Comment by kripken** [comment]
I think this is good to land without waiting for a test. But @junyuecao a test would be much appreciated as a followup! And would prevent this ever regressing.

**Pull request merged**